hail/python/hail/expr/functions.py

```diff
@@ -3603,16 +3617,31 @@ def sorted(collection,
     :class:`.ArrayExpression`
         Sorted array.
     """
-    ascending = ~reverse
+
+    rev = Env.get_uid()
+
+    def comp(l, r):
+        return If(IsNA(l),
+                  FalseIR(),
+                  If(IsNA(r),
+                     TrueIR(),
+                     If(Ref(rev),
+                        ApplyComparisonOp("GT", ApplyComparisonOp("Compare", l, r), I32(0)),
+                        ApplyComparisonOp("GT", ApplyComparisonOp("Compare", r, l), I32(0)))))
 
     if key is None:
-        return collection._method("sort", collection.dtype, ascending)
+        sorted = _sort_by(collection, lambda l, r: construct_expr(comp(l._ir, r._ir), tbool, collection._indices, collection._aggregations))
     else:
         with_key = collection.map(lambda elt: hl.tuple([key(elt), elt]))
-        ir = ArraySort(with_key._ir, ascending._ir, on_key=True)
-
-        indices, aggregations = unify_all(with_key, ascending)
-        return construct_expr(ir, with_key.dtype, indices, aggregations).map(lambda elt: elt[1])
+        sorted = _sort_by(
+            with_key,
+            lambda l, r: construct_expr(
+                comp(GetTupleElement(l._ir, 0), GetTupleElement(r._ir, 0)),
+                tbool,
+                with_key._indices,
+                with_key._aggregations)).map(lambda elt: elt[1])
+    indices, aggregations = unify_exprs(sorted, reverse)
+    return construct_expr(Let(rev, reverse._ir, sorted._ir), sorted.dtype, indices, aggregations)
```

Comment:
Actually, why write IR at all? We're in python after all. (Also, I think the missing logic is wrong when `rev=True`, can we add a test for that, I've fixed below)

```python
def lt(l, r):
    return (hl.case()
      .when(hl.is_missing(l), False)
      .when(hl.is_missing(r), True)
      .default(l < r)
def lt_rev(l, r, rev):
    return flip(rev, lt, l, r)
def flip(doflip, f, l, r):
     return hl.cond(doflip, f(r, l), f(l, r))
if key is None:
    return bind(rev => _sort_by(collection, lambda l, r: lt_rev(l, r, rev), reverse)
with_key = collection.map(lambda elt: hl.tuple([key(elt), elt]))
return bind(rev => _sort_by(with_key, lambda l, r: lt_rev(l[0], r[0], rev)), reverse).map(lambda e: e[1])
```